ql/termstructures/volatility/equityfx/andreasenhugevolatilityinterpl.cpp

```diff
@@ -0,0 +1,668 @@
+        const Handle<YieldTermStructure>& qTS,
+        InterpolationType interplationType,
+        CalibrationType calibrationType,
+        Size nGridPoints,
+        Real _minStrike,
+        Real _maxStrike,
+        const boost::shared_ptr<OptimizationMethod>& optimizationMethod,
+        const EndCriteria& endCriteria)
+    : calibrationSet_(calibrationSet),
+      spot_(spot),
+      rTS_(rTS),
+      qTS_(qTS),
+      interpolationType_(interplationType),
+      calibrationType_(calibrationType),
+      nGridPoints_(nGridPoints),
+      minStrike_(_minStrike),
+      maxStrike_(_maxStrike),
+      optimizationMethod_(optimizationMethod),
+      endCriteria_(endCriteria) {
+
+        QL_REQUIRE(nGridPoints > 2 && calibrationSet.size() > 0,
+                "undefined grid or calibration set");
+
+        std::set<Real> strikes;
+        std::set<Date> expiries;
+
+        for (Size i=0; i < calibrationSet.size(); ++i) {
+
+            const boost::shared_ptr<Exercise> exercise =
+                calibrationSet[i].first->exercise();
+
+            QL_REQUIRE(exercise->type() == Exercise::European,
+                    "European option required");
+
+            const Date expiry = exercise->lastDate();
+            expiries.insert(expiry);
+
+            const boost::shared_ptr<PlainVanillaPayoff> payoff =
+                boost::dynamic_pointer_cast<PlainVanillaPayoff>(
+                    calibrationSet[i].first->payoff());
+
+            QL_REQUIRE(payoff, "plain vanilla payoff required");
+
+            const Real strike = payoff->strike();
+            strikes.insert(strike);
+
+            registerWith(calibrationSet[i].second);
+        }
+
+        strikes_.assign(strikes.begin(), strikes.end());
+        expiries_.assign(expiries.begin(), expiries.end());
+
+        dT_.resize(expiries_.size());
+        expiryTimes_.resize(expiries_.size());
+
+        calibrationMatrix_ = std::vector< std::vector<Size> >(
+            expiries.size(), std::vector<Size>(strikes.size(), Null<Size>()));
+
+        for (Size i=0; i < calibrationSet.size(); ++i) {
+            const Date expiry =
+                calibrationSet[i].first->exercise()->lastDate();
+
+            const Size l = std::distance(expiries.begin(),
+                std::lower_bound(expiries.begin(), expiries.end(), expiry));
+
+            const Real strike =
+                boost::dynamic_pointer_cast<PlainVanillaPayoff>(
+                    calibrationSet[i].first->payoff())->strike();
+
+            const Size k = std::distance(strikes_.begin(),
+                std::find_if(strikes_.begin(), strikes_.end(),
+                    std::bind2nd(std::ptr_fun(close_enough), strike)));
+
+            calibrationMatrix_[l][k] = i;
+        }
+
+        registerWith(spot_);
+        registerWith(rTS_);
+        registerWith(qTS_);
+
+        update();
+    }
+
+    void AndreasenHugeVolatilityInterpl::update() {
+        LazyObject::update();
+
```

Comment:
The part below can probably go into `performCalculations`?

Reply:
IMO this would suppress the invocation of AndreasenHugeVolatilityInterpl::performCalculations in LazyObject::calculate after an invocation of AndreasenHugeVolatilityInterpl::update .

Reply:
Sorry, I'm not getting it.  May you elaborate?

Reply:
Sorry for the rather cryptic reply. 

A call of AndreasenHugeVolatilityInterpl::update should trigger a lazy revaluation of the volatility surface. Lazy here means that the revaluation should take place once with the first call of either

AndreasenHugeVolatilityInterpl::optionPrice , AndreasenHugeVolatilityInterpl::localVol or AndreasenHugeVolatilityInterpl::calibrationError

These methods aways call calculate(), which triggers performCalculations once if LazyObect::calculate_ is false.

I hope I haven't overseen something here.

Reply:
Yes, that will work.  You're calling `LazyObject::update()`, which sets `calculate_` to false.  What I was suggesting was to move the additional calculations in `update` at the beginning of `performCalculations` instead, so that they're done right before the revaluation of the surface instead of right away.  Or do you mean that you might want to calculate grid points etc. without also doing the revaluation?

Reply:
Sure, this can go to performCalculations. I'll change it.

Reply:
At which point `AndreasenHugeVolatilityInterpl::update` can just be inherited from `LazyObject`.